ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionHeartbeatService.java

```diff
@@ -0,0 +1,256 @@
+   * @param conf The {@link HiveConf} used to create the service. Used only during the firsst call
+   * @return Returns the singleton {@link CompactionHeartbeatService}
+   * @throws IllegalStateException Thrown when the service has already been shut down.
+   */
+  static CompactionHeartbeatService getInstance(HiveConf conf) {
+    if (INSTANCE == null) {
+      synchronized (CompactionHeartbeatService.class) {
+        if (INSTANCE == null) {
+          LOG.debug("Initializing compaction txn heartbeater service.");
+          INSTANCE = new CompactionHeartbeatService(conf);
+          Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+            try {
+              INSTANCE.shutdown();
+            } catch (InterruptedException e) {
+              Thread.currentThread().interrupt();
+            }
+          }));
+        }
+      }
+    }
+    if (INSTANCE.heartbeatExecutor.isShutdown()) {
+      throw new IllegalStateException("The CompactionHeartbeatService is already shut down!");
+    }
+    return INSTANCE;
+  }
+
+  private final ScheduledExecutorService heartbeatExecutor;
+  private final ObjectPool<IMetaStoreClient> clientPool;
+  private final long initialDelay;
+  private final long period;
+  private final ConcurrentHashMap<Long, TaskWrapper> tasks = new ConcurrentHashMap<>(30);
+
+  /**
+   * Starts the heartbeat for the given transaction
+   * @param txnId The id of the compaction txn
+   * @param lockId The id of the lock associated with the txn
+   * @param tableName Required for logging only
+   * @throws IllegalStateException Thrown when the heartbeat for the given txn has already been started.
+   */
+  void startHeartbeat(long txnId, long lockId, String tableName) {
+    if (tasks.containsKey(txnId)) {
+      throw new IllegalStateException("Heartbeat already started for TXN " + txnId);
+    }
+    LOG.info("Submitting heartbeat task for TXN {}", txnId);
+    CompactionHeartbeater heartbeater = new CompactionHeartbeater(txnId, lockId, tableName);
+    Future<?> submittedTask = heartbeatExecutor.scheduleAtFixedRate(heartbeater, initialDelay, period, TimeUnit.MILLISECONDS);
+    tasks.put(txnId, new TaskWrapper(heartbeater, submittedTask));
+  }
+
+  /**
+   * Stops the heartbeat for the given transaction
+   * @param txnId The id of the compaction txn
+   * @throws IllegalStateException Thrown when there is no {@link CompactionHeartbeater} task associated with the
+   * given txnId.
+   */
+  void stopHeartbeat(long txnId) {
+    LOG.info("Stopping heartbeat task for TXN {}", txnId);
+    TaskWrapper wrapper = tasks.get(txnId);
+    if (wrapper == null) {
+      throw new IllegalStateException("No registered heartbeat found for TXN " + txnId);
+    }
+    wrapper.future.cancel(false);
+    try {
+      wrapper.heartbeater.waitUntilFinish();
+    } catch (InterruptedException e) {
+      //Restore interrupted state, but let the compaction txn commit/abort
+      Thread.currentThread().interrupt();
```

Comment:
why do we need to handle InterruptedException here and not in upstream code?

Reply:
It can be handled in `CompactionTxn.close()` method as well. I thought it is more straightforward to handle it  here.